src/Symfony/Component/JsonPath/JsonCrawler.php

```diff
@@ -39,18 +39,33 @@ final class JsonCrawler implements JsonCrawlerInterface
     ];
 
     /**
-     * @param resource|string $raw
+     * @var resource|string|array|null
      */
-    public function __construct(
-        private readonly mixed $raw,
-    ) {
-        if (!\is_string($raw) && !\is_resource($raw)) {
-            throw new InvalidArgumentException(\sprintf('Expected string or resource, got "%s".', get_debug_type($raw)));
+    private mixed $raw = null;
+
+    /**
+     * @param resource|string|array $raw
+     *
+     * @throws \TypeError When the input is not a string or a resource
+     */
+    public function fromJson(mixed $raw): static
+    {
+        if (!\is_string($raw) && !\is_resource($raw) && !\is_array($raw)) {
+            throw new \TypeError(\sprintf('Expected string, array or resource, got "%s".', get_debug_type($raw)));
         }
+
+        $clone = clone $this;
+        $clone->raw = $raw;
+
+        return $clone;
     }
 
     public function find(string|JsonPath $query): array
     {
+        if (null === $this->raw) {
```

Comment:
let's add a constructor to not have to deal with this case?

Reply:
Maybe we can use `[]` as the default value of `$raw` instead of `null`: we wouldn't have to deal with this case and the crawler would be noop (i.e. always return `[]`) until one calls `fromJson()`. What do you think?

Reply:
Why not add a constructor? Related changes in the tests look like boilerplate we'd better remove to me.